lib/spack/docs/workflows.rst

```diff
@@ -0,0 +1,1150 @@
+-------------------
+
+``spack setup`` currently only supports CMake-based builds, in
+packages that subclass ``CMakePackage``.  The intent is that this
+mechanism should support a wider range of build systems; for example,
+GNU Autotools.  Someone well-versed in Autotools is needed to develop
+this patch and test it out.
+
+Python Distutils is another popular build system that should get
+``spack setup`` support.  For non-compiled languages like Python,
+``spack diy`` may be used.  Even better is to put the source directory
+directly in the user's ``PYTHONPATH``.  Then, edits in source files
+are immediately available to run without any install process at all!
+
+----------
+Conclusion
+----------
+
+The ``spack setup`` development workflow provides better automation,
+flexibility and safety than workflows relying on environment modules
+or filesystem views.  However, it has some drawbacks:
+
+#. It currently works only with projects that use the CMake build
+   system.  Support for other build systems is not hard to build, but
+   will require a small amount of effort for each build system to be
+   supported.  It might not work well with some IDEs.
+
+#. It only works with packages that sub-class ``StagedPackage``.
+   Currently, most Spack packages do not.  Converting them is not
+   hard; but must be done on a package-by-package basis.
+
+#. It requires that users are comfortable with Spack, as they
+   integrate Spack explicitly in their workflow.  Not all users are
+   willing to do this.
+
+==================
+Upstream Bug Fixes
+==================
+
+It is not uncommon to discover a bug in an upstream project while
+trying to build with Spack.  Typically, the bug is in a package that
+serves a dependency to something else.  This section describes
+procedure to work around and ultimately resolve these bugs, while not
+delaying the Spack user's main goal.
+
+^^^^^^^^^^^^^^^^^
+Buggy New Version
+^^^^^^^^^^^^^^^^^
+
+Sometimes, the old version of a package works fine, but a new version
+is buggy.  For example, it was once found that `Adios did not build
+with hdf5@1.10<https://github.com/LLNL/spack/issues/1683>_`.  If the
+old version of ``hdf5`` will work with ``adios``, the suggest
+procedure is:
+
+#. Revert ``adios`` to the old version of ``hdf5``.  Put in its
+   ``adios/package.py``:
+
+   .. code-block::
+       # Adios does not build with HDF5 1.10
+       # See: https://github.com/LLNL/spack/issues/1683
+       depends_on('hdf5@:1.9')
+
+#. Determine whether the problem is with ``hdf5`` or ``adios``, and
+   report the problem to the appropriate upstream project.  In this
+   case, the problem was with ``adios``.
+
+#. Once a new version of ``adios`` comes out with the bugfix, modify
+   ``adios/package.py`` to reflect it:
+
+       # Adios up to v1.10.0 does not build with HDF5 1.10
+       # See: https://github.com/LLNL/spack/issues/1683
+       depends_on('hdf5@:1.9', when='@:1.10.0')
+       depends_on('hdf5', when='@1.10.1:')
```

Comment:
Sorry, #1553 does not enlighten me.  Can you please be more specific on what the syntax should be?